cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/proto/FilterDeserializer.java

```diff
@@ -0,0 +1,93 @@
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+import java.lang.reflect.Type;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * JSON deserializer for {@link Filter}
+ */
+public class FilterDeserializer implements JsonDeserializer<Filter> {
+  private static final Type INT_RANGE_FILTER_TYPE =
+    new TypeToken<RangeFilter<Integer>>() { }.getType();
+  private static final Type LONG_RANGE_FILTER_TYPE =
+    new TypeToken<RangeFilter<Long>>() { }.getType();
+  private static final Type STRING_VALUE_FILTER_TYPE =
+    new TypeToken<ValueFilter<String>>() { }.getType();
+
+  /**
+   * Deserialize a JSON String as {@link Filter}. Determine the class and data type of
+   * the filter according to the field name that the filter contains.
+   */
+  @Nullable
+  @Override
+  public Filter deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+    throws JsonParseException {
+    if (json == null) {
+      return null;
+    }
+    if (!(json instanceof JsonObject)) {
+      throw new JsonParseException("Expected a JsonObject but found a " + json.getClass().getName());
+    }
+
+    JsonObject object = (JsonObject) json;
+    JsonElement fieldName = object.get("fieldName");
+    if (fieldName == null) {
+      throw new JsonParseException("Field name must be specified for filters");
+    }
+    ReportField field = ReportField.valueOfFieldName(fieldName.getAsString());
+    if (field == null) {
+      throw new JsonParseException(String.format("Invalid field name '%s'. Field name must be one of: [%s]", fieldName,
+                                                 String.join(", ", ReportField.FIELD_NAME_MAP.keySet())));
+    }
+    Filter filter = null;
+    // if the object contains "range" field, try to deserialize it as a range filter
+    if (object.get("range") != null) {
+      // Use the type token that matches the class of this field's value to deserialize the JSON
+      if (field.getValueClass().equals(Integer.class)) {
+        filter = context.deserialize(json, INT_RANGE_FILTER_TYPE);
+      } else if (field.getValueClass().equals(Long.class)) {
+        filter = context.deserialize(json, LONG_RANGE_FILTER_TYPE);
+      }
+      // otherwise, try to deserialize it as a value filter
+    } else if (field.getValueClass().equals(String.class)) {
+      // Use the type token that matches the class of this field's value to deserialize the JSON
+      filter = context.deserialize(json, STRING_VALUE_FILTER_TYPE);
+    }
+    if (filter == null) {
+      // this should never happen. If the field's applicable filters contains value filter,
+      // there must be a know class matches the class of its value
+      throw new JsonParseException(String.format("No applicable filter found for field %s with value type %s.",
+                                                 fieldName, field.getValueClass().getName()));
+    }
+    List<String> errors = filter.getErrors();
```

Comment:
is this needed, wouldn't errors in filter be caught during validation of request.